src/test/app/AMMClawback_test.cpp

```diff
@@ -0,0 +1,1444 @@
+            BEAST_EXPECT(
+                amm.expectLPTokens(carol, IOUAmount{1118033988749895, -12}));
+
+            env.require(balance(alice, gw["USD"](2000)));
+            env.require(balance(alice, gw2["EUR"](1000)));
+            env.require(balance(bob, gw["USD"](3000)));
+            env.require(balance(bob, gw2["EUR"](2500)));
+            env.require(balance(carol, gw["USD"](3000)));
+            env.require(balance(carol, gw2["EUR"](2750)));
+
+            // gw clawback all the bob's USD in amm. (2000 USD / 2500 EUR)
+            env(amm::ammClawback(gw, bob, USD, EUR, std::nullopt, std::nullopt),
+                ter(tesSUCCESS));
+            env.close();
+
+            BEAST_EXPECT(amm.expectBalances(
+                STAmount{USD, UINT64_C(4999999999999999), -12},
+                STAmount{EUR, UINT64_C(6249999999999999), -12},
+                IOUAmount{5590169943749475, -12}));
+
+            BEAST_EXPECT(
+                amm.expectLPTokens(alice, IOUAmount{4472135954999580, -12}));
+            BEAST_EXPECT(amm.expectLPTokens(bob, IOUAmount(0)));
+            BEAST_EXPECT(
+                amm.expectLPTokens(carol, IOUAmount{1118033988749895, -12}));
+
+            // Bob will get 2500 EUR back.
+            env.require(balance(alice, gw["USD"](2000)));
+            env.require(balance(alice, gw2["EUR"](1000)));
+            BEAST_EXPECT(
+                env.balance(bob, USD) ==
+                STAmount(USD, UINT64_C(3000000000000000), -12));
+
+            BEAST_EXPECT(
+                env.balance(bob, EUR) ==
+                STAmount(EUR, UINT64_C(5000000000000001), -12));
+            env.require(balance(carol, gw["USD"](3000)));
+            env.require(balance(carol, gw2["EUR"](2750)));
+
+            // gw2 clawback all carol's EUR in amm. (1000 USD / 1250 EUR)
+            env(amm::ammClawback(
+                    gw2, carol, EUR, USD, std::nullopt, std::nullopt),
+                ter(tesSUCCESS));
+            env.close();
+            BEAST_EXPECT(amm.expectBalances(
+                STAmount{USD, UINT64_C(3999999999999999), -12},
+                STAmount{EUR, UINT64_C(4999999999999999), -12},
+                IOUAmount{4472135954999580, -12}));
+
+            BEAST_EXPECT(
+                amm.expectLPTokens(alice, IOUAmount{4472135954999580, -12}));
+            BEAST_EXPECT(amm.expectLPTokens(bob, IOUAmount(0)));
+            BEAST_EXPECT(amm.expectLPTokens(carol, IOUAmount(0)));
+
+            // gw2 clawback all alice's EUR in amm. (4000 USD / 5000 EUR)
+            env(amm::ammClawback(
+                    gw2, alice, EUR, USD, std::nullopt, std::nullopt),
+                ter(tesSUCCESS));
+            env.close();
+
+            env.require(balance(carol, gw2["EUR"](2750)));
+            env.require(balance(carol, gw["USD"](4000)));
+            BEAST_EXPECT(!amm.ammExists());
+        }
+
+        // Test AMMClawback for USD/XRP pool. The assets are issued by different
```

Comment:
Typo? "The assets are issued by different issuer". It's XRP/USD.

Reply:
updated